File: Arrays 101/In Place Operations/Remove Duplicates From Sorted Array/RemoveDuplicatesfromSortedArray.py
# Given an integer array nums sorted in non-decreasing order, remove the duplicates in-place such that each unique element appears only once. The relative order of the elements should be kept the same.
# Since it is impossible to change the length of the array in some languages, you must instead have the result be placed in the first part of the array nums.
# More formally, if there are k elements after removing the duplicates, then the first k elements of nums should hold the final result. It does not matter what you leave beyond the first k elements.
# Return k after placing the final result in the first k slots of nums.
# Do not allocate extra space for another array. You must do this by modifying the input array in-place with O(1) extra memory.

class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        # input : int array
        # output: int (length of int array)

        # edge case: if length == 0 or 1
        if len(nums) == 0 or len(nums) == 1:
            return len(nums)

        # create pointer
        p = len(nums) - 2

        # iterate from the second last value
        while p >= 0 and len(nums) > 1:

            # if the previous element is the same as the current
            if nums[p] == nums[p+1]:

                # pop
                nums.pop(p)

                # decrement
                p -= 1

            # decrement
            else:
                p -= 1

        # return length of nums
        return len(nums)

File: Introduction to Data Structures: Array and String/5. Conclusion/Remove Duplicates from Sorted Array/RemoveDuplicatesFrom SortedArray.java
class Solution {
    public int removeDuplicates(int[] nums) {
                                  
        // Input: nums = [0,1,2,3,4,2,2,3,3,4]'
                                              
        // edge case
        if (nums.length == 0) {
            return 0;
        }
        
        if (nums.length == 1) {
            return 1;
        }
        
        int l = 1;
        int r = 1;
        
        while (r <= nums.length-1) {
            if (nums[r] != nums[r-1]) {
                nums[l] = nums[r];
                l += 1;
                r += 1;
            }
            else {
                r += 1;
            }
        }
        
        return l;
        
    }
}